cdap-ui/app/cdap/services/WizardStores/MicroserviceUpload/MicroserviceUploadStore.js

```diff
@@ -0,0 +1,349 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License s
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+import {combineReducers, createStore} from 'redux';
+import MicroserviceUploadActions from 'services/WizardStores/MicroserviceUpload/MicroserviceUploadActions';
+import MicroserviceUploadWizardConfig from 'services/WizardConfigs/MicroserviceUploadWizardConfig';
+import {getArtifactNameAndVersion} from 'services/helpers';
+import head from 'lodash/head';
+import cloneDeep from 'lodash/cloneDeep';
+import shortid from 'shortid';
+import T from 'i18n-react';
+
+// Defaults
+const defaultState = {
+  __complete: false,
+  __skipped: false,
+  __error: false
+};
+const defaultGeneralState = Object.assign({
+  instanceName: '',
+  description: '',
+  version: 1,
+  microserviceName: ''
+}, defaultState);
+
+const defaultJarState = Object.assign({}, {
+  contents: '',
+  fileMetadataObj: {}
+}, defaultState);
+
+const defaultJsonState = Object.assign({}, {
+  contents: '',
+  properties: {},
+  artifactExtends: '',
+  artifactPlugins: []
+}, defaultState);
+
+const defaultConfigureState = Object.assign({
+  instances: 1,
+  vcores: 1,
+  memory: 512,
+  ethreshold: 100
+}, defaultState, { __complete: true });
+
+const defaultEndpointsState = Object.assign({
+  fetch: 100,
+  in: [{
+    property: '',
+    uniqueId : shortid.generate()
+  }],
+  out: [{
+    property: '',
+    uniqueId : shortid.generate()
+  }],
+}, defaultState, { __complete: true });
+
+const defaultPropertiesState = Object.assign({
+  keyValues : {
+    pairs : [{
+      key : '',
+      value : '',
+      uniqueId : shortid.generate()
+    }]
+  }
+}, defaultState, { __complete: true });
+
+const defaultAction = {
+  type: '',
+  payload: {}
+};
+const defaultInitialState = {
+  general: defaultGeneralState,
+  uploadjar: defaultJarState,
+  uploadjson: defaultJsonState,
+  configure: defaultConfigureState,
+  endpoints: defaultEndpointsState,
+  properties: defaultPropertiesState
+};
+
+// Utilities. FIXME: Move to a common place?
+const isNil = (value) => value === null || typeof value === 'undefined' || value === '';
```

Comment:
Can't use lodash's `isEmpty`, because for some fields the default value is a number, and `isEmpty(<number>)` returns true. I have at least moved these checks to `services/helpers`, and use lodash's `isNil` instead of manually comparing to null/undefined ourselves.